src/debouncer.ts

```diff
@@ -0,0 +1,206 @@
+import { DBOSExecutor } from './dbos-executor';
+import {
+  getFunctionRegistration,
+  getFunctionRegistrationByName,
+  getRegisteredFunctionClassName,
+  UntypedAsyncFunction,
+} from './decorators';
+import { getDBOSErrorCode, QueueDedupIDDuplicated } from './error';
+import { DEBOUNCER_WORKLOW_NAME, INTERNAL_QUEUE_NAME } from './utils';
+
+// Parameters for the debouncer workflow
+interface DebouncerWorkflowParams {
+  workflowName: string;
+  workflowClassName: string;
+  startWorkflowParams?: StartWorkflowParams;
+  debounceTimeoutMs?: number;
+}
+
+// The message sent from a debounce to the debouncer workflow
+interface DebouncerMessage {
+  args: unknown[];
+  messageID: string;
+  debouncePeriodMs: number;
+}
+
+interface DebouncerConfig<Args extends unknown[], Return> {
+  workflow: (...args: Args) => Promise<Return>;
+  startWorkflowParams?: StartWorkflowParams;
+  debounceTimeoutMs?: number;
+}
+
+interface DebouncerClientConfig {
+  workflowName: string;
+  workflowClassName?: string;
+  startWorkflowParams?: StartWorkflowParams;
+  debounceTimeoutMs?: number;
+}
+
+const _DEBOUNCER_TOPIC = 'DEBOUNCER_TOPIC';
+
+export const debouncerWorkflowFunction = async (
+  initialDebouncePeriodMs: number,
+  cfg: DebouncerWorkflowParams,
+  ...args: unknown[]
+) => {
+  let workflowInputs = args;
+  const debounceDeadlineEpochMs = cfg.debounceTimeoutMs ? Date.now() + cfg.debounceTimeoutMs : Number.MAX_VALUE;
+  let debouncePeriodMs = initialDebouncePeriodMs;
+  while (Date.now() < debounceDeadlineEpochMs) {
+    const timeUntilDeadline = Math.max(debounceDeadlineEpochMs - Date.now(), 0);
+    const timeoutMs = Math.min(debouncePeriodMs, timeUntilDeadline);
+    const message = await DBOS.recv<DebouncerMessage>(_DEBOUNCER_TOPIC, timeoutMs / 1000);
+    if (message === null) {
+      break;
+    } else {
+      workflowInputs = message['args'];
+      debouncePeriodMs = message['debouncePeriodMs'];
+      await DBOS.setEvent(message.messageID, message.messageID);
+    }
+  }
+  const methReg = getFunctionRegistrationByName(cfg.workflowClassName, cfg.workflowName);
+  if (!methReg || !methReg.registeredFunction) {
+    throw Error(`Invalid workflow name provided to debouncer: ${cfg.workflowName}`);
+  }
+  const func = methReg?.registeredFunction as UntypedAsyncFunction;
+  await DBOS.startWorkflow(func, cfg.startWorkflowParams)(...workflowInputs);
+};
+
+export class Debouncer<Args extends unknown[], Return> {
+  private readonly cfg: DebouncerWorkflowParams;
+  constructor(params: DebouncerConfig<Args, Return>) {
+    const wInfo = getFunctionRegistration(params.workflow);
+    this.cfg = {
+      workflowName: wInfo?.name ?? params.workflow.name,
+      workflowClassName: getRegisteredFunctionClassName(params.workflow),
+      startWorkflowParams: params.startWorkflowParams,
+      debounceTimeoutMs: params.debounceTimeoutMs,
+    };
+  }
+
+  async debounce(debounceKey: string, debouncePeriodMs: number, ...args: Args) {
```

Comment:
Add a check that debouncerPeriodMs is a positive number?

Reply:
Done!